### PR TITLE
Remove the support for JPEG2000

### PR DESCRIPTION
**Pull request by shallawa**
#### 78cb9d96aa82cb8cfb1ecb6a1275efabfab31fab
<pre>
Remove the support for JPEG2000
<a href="https://bugs.webkit.org/show_bug.cgi?id=178758">https://bugs.webkit.org/show_bug.cgi?id=178758</a>
<a href="https://rdar.apple.com/35161822">rdar://35161822</a>

Reviewed by Michael Catanzaro.

Right now, no other browser is supporting JPEG2000. So there&apos;s no need for WebKit
to keep supporting it.

Clients of WebView and WKWebView can still force supporting JPEG2000 through the
private configuration/preference &apos;additionalSupportedImageTypes&apos;.

* LayoutTests/fast/images/image-formats-support-expected.txt:
* LayoutTests/fast/images/image-formats-support.html:
* LayoutTests/fast/images/resources/100x100-red-jp2-renamed.png: Added.
* LayoutTests/fast/images/resources/100x100-red.jp2: Added.
* Source/WebCore/platform/ImageDecoders.cmake:
* Source/WebCore/platform/MIMETypeRegistry.cpp:
* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::defaultSupportedImageTypes):
* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(WebCore::ScalableImageDecoder::create):
* Source/WebCore/platform/image-decoders/jpeg2000/JPEG2000ImageDecoder.cpp: Removed.
* Source/WebCore/platform/image-decoders/jpeg2000/JPEG2000ImageDecoder.h: Removed.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/100x100-red.jp2: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalSupportedImageTypes.mm:
(TEST):
* Tools/TestWebKitAPI/Tests/mac/AdditionalSupportedImageTypes.html:
* Tools/TestWebKitAPI/Tests/mac/AdditionalSupportedImageTypes.mm:
(TestWebKitAPI::runTest):
(TestWebKitAPI::TEST):
* Tools/gtk/jhbuild.modules:
* Tools/wpe/jhbuild.modules:

Canonical link: <a href="https://commits.webkit.org/273978@main">https://commits.webkit.org/273978@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/95cdb4dc849acc5877e0f208b0116983c02a8ba7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39872 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33294 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13372 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31733 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37902 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13666 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32802 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11897 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11886 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41135 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/31180 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33772 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33879 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37793 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/37027 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/12455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10024 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35940 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32828 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43920 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8434 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9013 "Found 156 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-dfg-eager-no-cjit, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-no-llint, stress/array-flatmap.js.dfg-eager-no-cjit-validate, wasm.yaml/wasm/gc/any.js.default-wasm, wasm.yaml/wasm/gc/any.js.wasm-bbq, wasm.yaml/wasm/gc/any.js.wasm-collect-continuously, wasm.yaml/wasm/gc/any.js.wasm-eager, wasm.yaml/wasm/gc/any.js.wasm-eager-jettison ... (failure)") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/cg/UTIRegistry.cpp [hunk 1]
```diff
@@ -50,7 +50,6 @@ const MemoryCompactLookupOnlyRobinHoodHashSet<String>& defaultSupportedImageType
             "public.png"_s,
             "public.tiff"_s,
 #if !PLATFORM(WIN)
-            "public.jpeg-2000"_s,
```
Comment (mcatanzaro) [inline Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:56]: This is enough to stop Safari from displaying JPEG 2000 images? If so, great. The Source/WebCore/platform/image-decoders/jpeg2000/ only needs to exist so long as Safari supports it.

**Review by mcatanzaro — APPROVED** [review]

**Comment by weinig** [comment]
My comment from 2017 still stands on this: https://bugs.webkit.org/show_bug.cgi?id=178758#c3. This should be a setting and something we do a linked-on-or-after with for API clients.

**Comment by grorg** [comment]
> My comment from 2017 still stands on this: https://bugs.webkit.org/show_bug.cgi?id=178758#c3. This should be a setting and something we do a linked-on-or-after with for API clients.

+1 to what @weinig says. Should be configurable, not just deleted.

**Review by aperezdc — COMMENTED** [review]
I'm very glad about this, thanks!

As for the WPE and GTK ports, I believe there is no need to make the JPEG2000 support configurable at runtime&mdash;we would rather see the code removed and the OpenJPEG dependency gone 💣

**Review by mcatanzaro — DISMISSED** [review]
> As for the WPE and GTK ports, I believe there is no need to make the JPEG2000 support configurable at runtime—we would rather see the code removed and the OpenJPEG dependency gone bomb

For sure, but that should not happen until _after_ the Apple code is removed. So if Apple isn't willing to remove this yet, we need to keep it too, I'm afraid. Have to remain web-compatible with Safari.

**Comment by aperezdc** [comment]
> > As for the WPE and GTK ports, I believe there is no need to make the JPEG2000 support configurable at runtime—we would rather see the code removed and the OpenJPEG dependency gone bomb
> 
> For sure, but that should not happen until _after_ the Apple code is removed. So if Apple isn't willing to remove this yet, we need to keep it too, I'm afraid. Have to remain web-compatible with Safari.

My understanding from @weinig's comment above is that in the Apple ports the JPEG2000 support would remain enabled for applications which link against older versions of the WebKit framework and that thereforee would expect that the format is supported, but that Safari (and newer applications) would have the JPEG2000 support disabled&mdash;which I think is fine and already would push towards websites stopping serving images in JPEG2000 format.

There would be still the question of when we would remove JPEG2000 support in the GTK+WPE ports. I think we may want to keep it in the 2.38.x series in order to give time for versions of Safari that won't support JPEG2000 to come out. This means that we might want to still fix related issues in our 2.38 release branch. Opionions, @mcatanzaro and @carlosgcampos?

**Comment by mcatanzaro** [comment]
If it's only a linked-on-or-after check, then yes, it's safe for us to delete the JPEG 2000 image decoder. If there will be a preference to toggle JPEG 2000 support, then no, we should maintain the image decoder at least until we see a released version of Safari that doesn't support JPEG 2000.

**Comment by shallawa** [comment]
> > My comment from 2017 still stands on this: https://bugs.webkit.org/show_bug.cgi?id=178758#c3. This should be a setting and something we do a linked-on-or-after with for API clients.
> 
> +1 to what @weinig says. Should be configurable, not just deleted.

To enable unsupported images like HEIF and HEIC on macOS/iOS we do not provide UI to enable them. But we provide the private configuration/preference additionalSupportedImageTypes. Similarly you can do the same for JPEG2000 after landing this patch:

WK1:

```
RetainPtr<WebPreferences> preferences = adoptNS([[WebPreferences alloc] initWithIdentifier:nil]);
[preferences setAdditionalSupportedImageTypes:@[@"public.jpeg-2000"]];

RetainPtr<WebView> webView = adoptNS([[WebView alloc] initWithFrame:NSMakeRect(0, 0, 120, 200) frameName:nil groupName:nil]);
[webView setPreferences:preferences.get()];

```

WK2:

```
RetainPtr<WKWebViewConfiguration> configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
[configuration _setAdditionalSupportedImageTypes:@[@"public.jpeg-2000"]];
RetainPtr<WKWebView> webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);

```

For the WPE and GTK ports, unfortunately this preference is not available.

**Comment by aperezdc** [comment]
> For the WPE and GTK ports, unfortunately this preference is not available.

This is not a problem for us; we only added JPEG2000 support in order to make some websites work due to servers sending pictures in that format because of our UA string being similar to Safari's. If Safari stops supporting the format, web sites will also stop sending pictures in that format to applications using the GTK and WPE ports. As for applications embedding web views to load their custom web content, I highly doubt they would use JPEG2000.

TL;DR: This patch works for GTK/WPE as-is 👍🏼

**Comment by mcatanzaro** [comment]
> TL;DR: This patch works for GTK/WPE as-is 👍🏼

Assuming the next Safari will not use this preference and not enable JPEG 2000, yes. If so, great. But if this assumption is not correct, then please do not remove the JPEG 2000 decoder.

**Comment by shallawa** [comment]
> > TL;DR: This patch works for GTK/WPE as-is 👍🏼
> 
> Assuming the next Safari will not use this preference and not enable JPEG 2000, yes. If so, great. But if this assumption is not correct, then please do not remove the JPEG 2000 decoder.

I think the plan for Safari is to remove the support for JPEG 2000. I think the private configuration/preference is enough as a way to force enabling them back for Apple ports.

For GTK/WPE, if you feel more comfortable, I can leave the decoder files as they are. We can delete them later in a separate patch.

**Comment by shallawa** [comment]
> My comment from 2017 still stands on this: https://bugs.webkit.org/show_bug.cgi?id=178758#c3. This should be a setting and something we do a linked-on-or-after with for API clients.

I can add a feature control flag if the private configuration/preference is not discoverable or we want give the option to enable it in Safari. Do you think we should do that?

**Comment by mcatanzaro** [comment]
> I think the plan for Safari is to remove the support for JPEG 2000.

OK, great.

> For GTK/WPE, if you feel more comfortable, I can leave the decoder files as they are. We can delete them later in a separate patch.

In that case, you can go ahead and delete them now. The change won't reach the next stable branch until next March, anyway. If you change your mind about removing it from Safari and we encounter further web compat issues, we can always bring it back.

(BTW what we care about is Safari's default behavior. If it maintains an option to enable JPEG 2000 that is off by default, that's fine: we don't care.)

**Comment by mcatanzaro** [comment]
> I think the plan for Safari is to remove the support for JPEG 2000. I think the private configuration/preference is enough as a way to force enabling them back for Apple ports.

Would be real cool to see this go through.

**Comment by mcatanzaro** [comment]
Any plans to add the linked-on-or-after check requested by @weinig?

**Comment by shallawa** [comment]
The plan is to remove the support for JPEG2000 from all ports. For apple ports, there an API to add the support for WebView and WKWebView if needed.

**Review by shallawa — COMMENTED** [review]
The plan is to remove the support for JPEG2000 from all ports. For apple ports, there an API to add the support for WebView and WKWebView if needed.

**Review by mcatanzaro — APPROVED** [review]
